### PR TITLE
[Full-powered Tokenfactory] Apply `TrackBeforeSend`, `BlockBeforeSend`

### PR DESCRIPTION
**Pull request by mattverse**
Closes: #2567, #2570, #2629

## What is the purpose of the change
(Note that is PR'd against `fullpowere-tokenfactory` branch, not the main branch.

cref: https://github.com/osmosis-labs/cosmos-sdk/pull/334
Having separated out `BeforeSend` hook into two subsequent hooks `TrackBeforSend`, `BlockBeforeSend`, full powered token factory also has to implement the new hooks that has been changed. Both methods / hooks call the before send listener, the difference being `BlockBeforeSend` catches error returning from the listener, while `TrackBeforeSend` does not catch or return errors that are returned from the contract

### Minor changes included in this PR 
- Changed `go.mod` so that the cosmos-sdk dependecy points to  the commit in https://github.com/osmosis-labs/cosmos-sdk/pull/334 (this is what also caused all the changes in proto files) 
- Changed `SetBeforeSendHook` msg to `SetBeforeSendListener`. Changed method names from `setBeforeSendHook` to `SetBefo
- Changed so that `PerformBurn` does not check for burn address

### CODE REVIEW

File: proto/buf.lock [hunk 1]
```diff
@@ -8,12 +8,12 @@ deps:
   - remote: buf.build
     owner: cosmos
     repository: cosmos-sdk
-    commit: 3db8b415c96b4b0d85ec0a54ac5701d3
+    commit: 1d1523497b394c50ac17605136043cbc
```
Comment (mattverse) [inline proto/buf.lock:11]: These are buf mod updates that has been made due to cosmos sdk dependency update

File: simulation/simtypes/legacy.go [hunk 2]
```diff
@@ -71,10 +71,10 @@ func GenAndDeliverTx(
...
 	}
 
-	_, _, err = app.Deliver(txGen.TxEncoder(), tx)
+	gasInfo, _, err := app.Deliver(txGen.TxEncoder(), tx)
 	if err != nil {
 		return simulation.NoOpMsg(moduleName, msg.Type(), "unable to deliver tx"), nil, err
 	}
 
-	return simulation.NewOperationMsg(msg, true, "", nil), nil, nil
+	return simulation.NewOperationMsg(msg, true, "", gasInfo.GasWanted, gasInfo.GasUsed, nil), nil, nil
```
Comment (mattverse) [inline simulation/simtypes/legacy.go:79]: Changes made due to dependency update

File: wasmbinding/message_plugin.go [hunk 3]
```diff
@@ -184,9 +184,6 @@ func PerformBurn(f *tokenfactorykeeper.Keeper, ctx sdk.Context, contractAddr sdk
 	if burn == nil {
 		return wasmvmtypes.InvalidRequest{Err: "burn token null mint"}
 	}
-	if burn.BurnFromAddress != "" && burn.BurnFromAddress != contractAddr.String() {
-		return wasmvmtypes.InvalidRequest{Err: "BurnFromAddress must be \"\""}
-	}
```
Comment (mattverse) [inline wasmbinding/message_plugin.go:187]: Confirm we want to allow burn from any address if msg sender is denom admin

File: x/mint/keeper/keeper_test.go [hunk 4]
```diff
@@ -351,50 +351,52 @@ func (suite *KeeperTestSuite) TestCreateDeveloperVestingModuleAccount() {
...
-				suite.Require().Equal(supplyOffsetBefore, bankKeeper.GetSupplyOffset(ctx, sdk.DefaultBondDenom))
-				return
-			}
-			suite.Require().NoError(actualError)
-			suite.Require().Equal(supplyWithOffsetBefore.Amount.Sub(sdk.NewInt(keeper.DeveloperVestingAmount)), bankKeeper.GetSupplyWithOffset(ctx, sdk.DefaultBondDenom).Amount)
-			suite.Require().Equal(supplyOffsetBefore.Sub(sdk.NewInt(keeper.DeveloperVestingAmount)), bankKeeper.GetSupplyOffset(ctx, sdk.DefaultBondDenom))
-		})
-	}
-}
+// func (suite *KeeperTestSuite) TestSetInitialSupplyOffsetDuringMigration() {
```
Comment (mattverse) [inline x/mint/keeper/keeper_test.go:354]: This should be removed once we merge with main

File: wasmbinding/message_plugin.go (hunk 3 above)
Comment (ValarDragon) [reply wasmbinding/message_plugin.go:187]: Yes, except if burn from address is a module account

**Review by ValarDragon — APPROVED** [review]
LGTM! Just two notes:
* Block Burn from module acc
* Make SudoMsg into two different message types (you pointed this out!)

Awesome job

File: proto/osmosis/tokenfactory/v1beta1/query.proto [hunk 5]
```diff
@@ -33,12 +33,12 @@ service Query {
         "/osmosis/tokenfactory/v1beta1/denoms_from_creator/{creator}";
   }
 
-    // DenomsFromCreator defines a gRPC query method for fetching all
+  // DenomsFromCreator defines a gRPC query method for fetching all
```
Comment (czarcas7ic) [inline proto/osmosis/tokenfactory/v1beta1/query.proto:36]:
```suggestion
  // DenomBeforeSendListener defines a gRPC query method for fetching all
```

File: proto/osmosis/tokenfactory/v1beta1/query.proto [hunk 6]
```diff
@@ -33,12 +33,12 @@ service Query {
...
   // denominations created by a specific admin/creator.
-  rpc DenomBeforeSendHook(QueryDenomBeforeSendHookRequest)
-      returns (QueryDenomBeforeSendHookResponse) {
+  rpc DenomBeforeSendListener(QueryDenomBeforeSendListenerRequest)
+      returns (QueryDenomBeforeSendListenerResponse) {
     option (google.api.http).get =
-    "/osmosis/tokenfactory/v1beta1/denoms/{denom}/before_send_hook";
+        "/osmosis/tokenfactory/v1beta1/denoms/{denom}/before_send_hook";
```
Comment (czarcas7ic) [inline proto/osmosis/tokenfactory/v1beta1/query.proto:41]: It seems we are calling this a "before send listener" but it retains the "before_send_hook" nomenclature. Is this intentional?

File: wasmbinding/test/messages_test.go [hunk 7]
```diff
@@ -336,7 +336,7 @@ func TestBurn(t *testing.T) {
 				Amount:          mintAmount,
 				BurnFromAddress: lucky.String(),
 			},
-			expErr: true,
+			expErr: false,
```
Comment (czarcas7ic) [inline wasmbinding/test/messages_test.go:339]: Not quite understanding this one. Why do we not expect an error when burning from an address that is not the denom admin?

File: x/tokenfactory/client/cli/query.go [hunk 8]
```diff
@@ -122,7 +122,7 @@ func GetCmdDenomsFromCreator() *cobra.Command {
 }
 
 // GetCmdDenomAuthorityMetadata returns the authority metadata for a queried denom
```
Comment (czarcas7ic) [inline x/tokenfactory/client/cli/query.go:124]:
```suggestion
// GetCmdDenomBeforeSendListner returns the authority metadata for a queried denom
```

File: x/tokenfactory/client/cli/query.go [hunk 9]
```diff
@@ -122,7 +122,7 @@ func GetCmdDenomsFromCreator() *cobra.Command {
...
-func GetCmdDenomBeforeSendHook() *cobra.Command {
+func GetCmdDenomBeforeSendListner() *cobra.Command {
```
Comment (czarcas7ic) [inline x/tokenfactory/client/cli/query.go:125]: Same question as before, are we intentionally keeping the below "denom-before-send-hook" nomenclature despite changing it to "before send listener"?

File: x/tokenfactory/client/cli/tx.go [hunk 10]
```diff
@@ -190,7 +190,7 @@ func NewChangeAdminCmd() *cobra.Command {
 }
 
 // NewChangeAdminCmd broadcast MsgChangeAdmin
```
Comment (czarcas7ic) [inline x/tokenfactory/client/cli/tx.go:192]:
```suggestion
// NewSetBeforeSendListenerCmd broadcast MsgChangeAdmin
```

File: x/tokenfactory/client/cli/tx.go [hunk 11]
```diff
@@ -190,7 +190,7 @@ func NewChangeAdminCmd() *cobra.Command {
...
-func NewSetBeforeSendHookCmd() *cobra.Command {
+func NewSetBeforeSendListenerCmd() *cobra.Command {
```
Comment (czarcas7ic) [inline x/tokenfactory/client/cli/tx.go:193]: Same question

File: x/tokenfactory/types/keys.go [hunk 12]
```diff
@@ -25,11 +25,11 @@ const (
...
-	DenomAuthorityMetadataKey      = "authoritymetadata"
-	DenomsPrefixKey                = "denoms"
-	CreatorPrefixKey               = "creator"
-	AdminPrefixKey                 = "admin"
-	BeforeSendHookAddressPrefixKey = "beforesendhook"
+	DenomAuthorityMetadataKey          = "authoritymetadata"
+	DenomsPrefixKey                    = "denoms"
+	CreatorPrefixKey                   = "creator"
+	AdminPrefixKey                     = "admin"
+	BeforeSendListenerAddressPrefixKey = "beforesendlistner"
```
Comment (czarcas7ic) [inline x/tokenfactory/types/keys.go:32]: listener is misspelled here

File: x/tokenfactory/types/msgs.go [hunk 13]
```diff
@@ -8,13 +8,13 @@ import (
...
-	TypeMsgChangeAdmin       = "change_admin"
-	TypeMsgSetDenomMetadata  = "set_denom_metadata"
-	TypeMsgSetBeforeSendHook = "set_before_send_hook"
+	TypeMsgCreateDenom           = "create_denom"
+	TypeMsgMint                  = "mint"
+	TypeMsgBurn                  = "burn"
+	TypeMsgForceTransfer         = "force_transfer"
+	TypeMsgChangeAdmin           = "change_admin"
+	TypeMsgSetDenomMetadata      = "set_denom_metadata"
+	TypeMsgSetBeforeSendListener = "set_before_send_listner"
```
Comment (czarcas7ic) [inline x/tokenfactory/types/msgs.go:17]: listener msg is misspelled here

**Review by czarcas7ic — APPROVED** [review]
Nice work! 

I am still unconvinced that we can't just check addresses in the original before send hook and provide the respective logic on the module accounts only, but I don't feel strongly about this.

File: x/tokenfactory/client/cli/query.go (hunk 9 above)
Comment (mattverse) [reply x/tokenfactory/client/cli/query.go:125]: Ah good catch! Fixed!

File: x/tokenfactory/client/cli/tx.go (hunk 11 above)
Comment (mattverse) [reply x/tokenfactory/client/cli/tx.go:193]: great catch!

File: proto/osmosis/tokenfactory/v1beta1/query.proto (hunk 6 above)
Comment (mattverse) [reply proto/osmosis/tokenfactory/v1beta1/query.proto:41]: not intentional 🥲

File: wasmbinding/test/messages_test.go (hunk 7 above)
Comment (mattverse) [reply wasmbinding/test/messages_test.go:339]:
```
if burn.BurnFromAddress != "" && burn.BurnFromAddress != contractAddr.String() {
		return wasmvmtypes.InvalidRequest{Err: "BurnFromAddress must be \"\""}
	}
```

Previously, we had a check for dis-allowing admin accounts to burn from all accounts, this change removes that constraint

File: wasmbinding/message_plugin.go (hunk 3 above)
Comment (mattverse) [reply wasmbinding/message_plugin.go:187]: creating this logic in the msg server layer

File: wasmbinding/test/messages_test.go (hunk 7 above)
Comment (czarcas7ic) [reply wasmbinding/test/messages_test.go:339]: Oh wow I had a fundamental incorrect understanding of what burn from meant. I thought it meant "burn tokens AS this user" not "burn tokens RESIDING in this users account". Thanks for the clarification!

File: x/tokenfactory/keeper/msg_server.go [hunk 14]
```diff
@@ -95,6 +97,12 @@ func (server msgServer) Burn(goCtx context.Context, msg *types.MsgBurn) (*types.
 		msg.BurnFromAddress = msg.Sender
 	}
 
+	accountI := server.Keeper.accountKeeper.GetAccount(ctx, sdk.AccAddress(msg.BurnFromAddress))
+	_, ok := accountI.(authtypes.ModuleAccountI)
+	if ok {
+		return nil, types.ErrBurnFromModuleAccount
+	}
+
```
Comment (mattverse) [inline x/tokenfactory/keeper/msg_server.go:105]: Can someone confirm this newly added logic for checking if the burn from account is not an module account before merging? @ValarDragon @czarcas7ic

Comment (czarcas7ic) [reply x/tokenfactory/keeper/msg_server.go:105]: This make sense to me, but is there any reason tokenfactory does not have a msg_server_test? It appears other modules have this

Comment (mattverse) [reply x/tokenfactory/keeper/msg_server.go:105]: @p0mvn Would you be able to check this one logic? Thanks

**Comment by mattverse** [comment]
Status of this PR is that I need approval on the latest commit on adding a check for module account before merging!

**Pull request closed without merging**